File: test/cpu/chk.cc
#define BOOST_TEST_DYN_LINK
#include "68040.hpp"

#include "test.hpp"
#include <boost/test/data/monomorphic.hpp>
#include <boost/test/data/test_case.hpp>
#include <boost/test/unit_test.hpp>
namespace bdata = boost::unit_test::data;
struct F_CHK {
    F_CHK() {
        // CHK.W %D2, %D3
        TEST::SET_W(0, 0040600 | 3 << 9 | 2);
        TEST::SET_W(2, TEST_BREAK);

        // CHK.L %D2, %D3
        TEST::SET_W(4, 0040400 | 3 << 9 | 2);
        TEST::SET_W(6, TEST_BREAK);
        jit_compile(0, 8);
    }};
BOOST_FIXTURE_TEST_SUITE(CHK, Prepare, *boost::unit_test::fixture<F_CHK>())


BOOST_AUTO_TEST_SUITE(Word)
BOOST_AUTO_TEST_CASE(In) {
    cpu.D[3] = 1000;
    cpu.D[2] = 2000;
    run_test(0);
    BOOST_TEST(cpu.ex_n == EXCEPTION_NUMBER::NO_ERR);
}

BOOST_AUTO_TEST_CASE(Under) {
    cpu.D[3] = -300;
    cpu.D[2] = 2000;
    run_test(0);
    BOOST_TEST(cpu.ex_n == EXCEPTION_NUMBER::CHK_FAIL);
    BOOST_TEST(cpu.N);
}

BOOST_AUTO_TEST_CASE(Over) {
    TEST::SET_W(0, 0040600 | 3 << 9 | 2);
    cpu.D[3] = 2300;
    cpu.D[2] = 2000;
    run_test(0);
    BOOST_TEST(cpu.ex_n == EXCEPTION_NUMBER::CHK_FAIL);
    BOOST_TEST(!cpu.N);
}
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE(Long)
BOOST_AUTO_TEST_CASE(In) {
    cpu.D[3] = 100000;
    cpu.D[2] = 200000;
    run_test(4);
    BOOST_TEST(cpu.ex_n == EXCEPTION_NUMBER::NO_ERR);
}

BOOST_AUTO_TEST_CASE(Under) {
    cpu.D[3] = -30000;
    cpu.D[2] = 200000;
    run_test(4);
    BOOST_TEST(cpu.ex_n == EXCEPTION_NUMBER::CHK_FAIL);
    BOOST_TEST(cpu.N);
}

BOOST_AUTO_TEST_CASE(Over) {
    cpu.D[3] = 230000;
    cpu.D[2] = 200000;
    run_test(4);
    BOOST_TEST(cpu.ex_n == EXCEPTION_NUMBER::CHK_FAIL);
    BOOST_TEST(!cpu.N);
}
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE_END()

File: test/lua/cpu/chk.lua
-- CHK.W %D1, %D4
MEM[0].W = OCT"044601"
MEM[2].W = TEST.BREAK

-- CHK.L %D2, %D4
MEM[4].W = OCT"044402"
MEM[6].W = TEST.BREAK

JIT_COMPILE(0, 8)

local run = {}
run.W = {
   inRage = function()
      CPU.D[5].H = 2000
      CPU.D[2].H = 1000
      TEST.RUN(0)
   end,
   under = function()
      CPU.D[5].H = -300
      CPU.D[2].H = 2000
      TEST.RUN(0, "chk failure")
      TEST.CHECK("CPU.CC.N", true)
   end,
   over = function()
      CPU.D[5].H = 2300
      CPU.D[2].H = 2000
      TEST.RUN(0, "chk failure")
      TEST.CHECK("CPU.CC.N", false)
   end,
}

run.L = {
   inRage = function()
      CPU.D[5].I = 200000
      CPU.D[3].I = 100000
      TEST.RUN(4)
   end,
   under = function()
      CPU.D[5].I = -30000
      CPU.D[3].I = 200000
      TEST.RUN(4, "chk failure")
      TEST.CHECK("CPU.CC.N", true)
   end,
   over = function()
      CPU.D[5].I = 230000
      CPU.D[3].I = 200000
      TEST.RUN(4, "chk failure")
      TEST.CHECK("CPU.CC.N", false)
   end,
}

return run
